R/Spectra-functions.R

```diff
@@ -153,3 +153,109 @@ applyProcessing <- function(object, f = dataStorage(object),
         FALSE
     } else TRUE
 }
+
+#' @title Spectrum comparison
+#'
+#' @description
+#'
+#' Compare each spectrum in `x` with each spectrum in `y` with function `FUN`.
+#' Mapping between the peaks in both spectra can be defined with the `MAPFUN`
+#' function.
+#'
+#' @note
+#'
+#' Results might be slightly different if `x` and `y` are switched, because of
+#' the matching of the peaks by their m/z between spectra. Matching is performed
+#' by applying a `tolerance` and `ppm` to the second spectrum (i.e. from `y`)
+#' and, especially for `ppm`, the maximal accepted difference depend thus on
+#' the m/z values from the second spectrum.
+#'
+#' @param x [Spectra()]
+#'
+#' @param y [Spectra()]
+#'
+#' @param FUN `function` to compare the (m/z matched) intensities from each
+#'     spectrum in `x` with those in `y`.
+#'
+#' @param MAPFUN `function` to map peaks between the two compared spectra. See
+#'     [joinPeaks()].
+#'
+#' @param tolerance `numeric(1)` allowing to define a constant maximal accepted
+#'     difference between m/z values for peaks to be matched.
+#'
+#' @param ppm `numeric(1)` allowing to define a relative, m/z-dependent,
+#'     maximal accepted difference between m/z values for peaks to be matched.
+#'
+#' @param ... additional parameters passed to `FUN` and `MAPFUN`.
+#'
+#' @return
+#'
+#' `matrix` with number of rows equal to length of `x` and number of columns
+#' equal `length(y)`.
+#'
+#' @importFrom stats cor
+#'
+#' @examples
+#'
+#' library(Spectra)
+#' fl <- system.file("TripleTOF-SWATH/PestMix1_SWATH.mzML", package = "msdata")
+#' sps <- Spectra(fl, source = MsBackendMzR())
+#'
+#' sps <- sps[1:10]
+#' res <- .compare_spectra(sps, sps, ppm = 20)
+#' res <- .compare_spectra(sps, sps, method = "spearman", ppm = 40)
+#'
+#' res <- .compare_spectra(sps[1], sps)
+#' res <- .compare_spectra(sps, sps[1])
+#' @noRd
+.compare_spectra <- function(x, y = NULL, MAPFUN = joinPeaks, tolerance = 0,
+                             ppm = 20, FUN = cor, ...) {
+    x_idx <- seq_along(x)
+    y_idx <- seq_along(y)
+    mat <- matrix(NA_real_, nrow = length(x_idx), ncol = length(y_idx),
+                  dimnames = list(spectraNames(x), spectraNames(y)))
+    ## Might need some tuning - bplapply?
+    for (i in x_idx) {
+        for (j in y_idx) {
+            peak_map <- MAPFUN(peaks(x[i])[[1]], peaks(y[j])[[1]],
+                               tolerance = tolerance, ppm = ppm, ...)
+            mat[i, j] <- FUN(peak_map$x[, 2], peak_map$y[, 2], ...)
```

Comment:
It is not a big deal but `peak_map[[1L]][, 2L]` would be slightly faster than `peak_map$x[, 2]`:
```r
library("microbenchmark")
l <- list(x = 1:1e4, y = 1:1e4)
microbenchmark(l$x, l[[1L]], l$y, l[[2L]])
Unit: nanoseconds
    expr min    lq   mean median    uq  max neval cld
     l$x 217 231.0 355.83  249.5 268.5 9889   100   a
 l[[1L]] 157 172.0 206.36  178.5 188.0 2669   100   a
     l$y 230 247.5 280.07  267.5 285.0  912   100   a
 l[[2L]] 150 165.5 182.89  177.5 197.5  304   100   a
```

Reply:
OK, I will change it.